scripts/migration/migrate_root_and_parent_on_node.py

```diff
@@ -0,0 +1,79 @@
+"""
+This will add a parent field and a parent_node field to all nodes.
+Root  will be the primary key of the originating parent node.
+Parent_node will be the first primary parent
+Done so that you can filter on both root nodes and parent nodes with a DB query
+
+"""
+
+import sys
+import logging
+from modularodm import Q
+from website import models
+from website.app import init_app
+from scripts import utils as script_utils
+from framework.transactions.context import TokuTransaction
+
+logging.basicConfig(level=logging.INFO)
+logger = logging.getLogger(__name__)
+
+
+def do_migration():
+    init_app(routes=False)
+    logger.warn('parent and root fields will be added to all nodes.')
+    all_nodes = models.Node.find(Q('is_deleted', 'eq', False) & Q('root', 'eq', None))
+    all_nodes_count = all_nodes.count()
+    touched_counter = 0
+    logger.info('There are {} total nodes'.format(all_nodes_count))
+    for node in all_nodes:
+        if not getattr(node, '_parent_node', None):
+            touched_counter += 1
+            node.save()
+            children = [child for child in node.get_descendants_recursive(include=lambda n: n.primary and not n.is_deleted)]
+            logger.info(
+                '{}/{}: touched node {} with children {}'.format(
+                    touched_counter,
+                    all_nodes_count,
+                    node._id,
+                    [child._id for child in children]
+                )
+            )
+
+            assert node.root._id == node._id
+            assert not getattr(node, 'parent_node', None)
+            logger.info('Parent Node Saving: Saved Node {} with root {}'.format(node._id, node.root))
+
+            for child in children:
+                touched_counter += 1
+                logger.info(
+                    '{}/{}: touched node {} with parent {}'.format(
+                        touched_counter,
+                        all_nodes_count,
+                        child._id,
+                        child.parent_id
+                    )
+                )
+
+                child.save()
+                logger.info('The child root id is {} and the node id is {}'.format(child.root._id, node._id))
+                logger.info('Child Node saved: Verifying that save Node {} with root {}'.format(child._id, child.root))
+                assert child.parent_node._id == child.parent_id
+                assert child.root._id == node._id
+
+    assert all_nodes_count == touched_counter
+
+
+def main(dry=True):
+    init_app(set_backends=True, routes=False)  # Sets the storage backends on all models
+    with TokuTransaction():
```

Comment:
Since the whole migration is wrapped in a transaction, it won't actually be resume-able. We can fix this by only transacting updates to individual nodes.